File: frontend/src/contracts/systemStatusContract.ts
/**
 * System-Status-Contract — Zod-Spiegel zu /api/status.
 *
 * Backend: backend/app/api/status.py::get_status
 * passthrough() auf allen Unterobjekten, damit Backend additive Felder
 * (z. B. neue gpu/disk-Subkeys) liefern darf ohne Frontend-Drift-Alarm.
 */
import { z } from 'zod'

/**
 * Spiegel von `backend/app/contracts/system_status_contract.py::StatusCheckError`.
 *
 * Ersetzt seit #1458 den rohen `str(exc)`, der zuvor an drei Stellen
 * (neo4j, ollama, disk) direkt in die Antwort floss — ein Informationsleck
 * (Pfade, Hostnamen, Treiberdetails) und für die UI unrenderbarer
 * Traceback-Text. `code` ist bewusst `z.string()` statt eines Enums: ein
 * neuer Backend-Code darf den Zod-Parse nicht invalidieren, das Frontend
 * übersetzt ihn mit Fallback für unbekannte Werte
 * (`unreachable` | `timeout` | `auth` | `unexpected` sind die bekannten
 * Werte, siehe `StatusErrorCode` im Backend-Contract).
 */
export const StatusCheckErrorSchema = z
  .object({
    code: z.string(),
  })
  .passthrough()

export const SystemStatusBackendSchema = z
  .object({
    ok: z.boolean(),
    version: z.string().optional(),
    auth_mode: z.string().optional(),
  })
  .passthrough()

export const SystemStatusNeo4jSchema = z
  .object({
    reachable: z.boolean(),
    error: StatusCheckErrorSchema.nullable().optional(),
    uri: z.string().optional(),
    is_connected: z.boolean().optional(),
    last_success_ts: z.string().nullable().optional(),
  })
  .passthrough()

/**
 * Spiegel von `backend/app/contracts/system_status_contract.py::SystemStatusOllama`
 * (generiert nach `schemas/system-status-ollama.schema.json`).
 *
 * `reachable` ist dreiwertig: `true` erreichbar, `false` Probe fehlgeschlagen,
 * `null` Probe übersprungen — Letzteres ist KEIN Fehlerzustand und darf nicht
 * als "offline" gerendert werden.
 */
export const SystemStatusOllamaSchema = z
  .object({
    reachable: z.boolean().nullable(),
    skipped: z.boolean().optional(),
    // Maschinenlesbarer i18n-Schlüssel. `reason` ist reines Debug-Feld und
    // gehört nicht in die UI — es enthält englische Backend-Prosa.
    skipped_provider: z.string().nullable().optional(),
    reason: z.string().nullable().optional(),
    base_url: z.string().nullable().optional(),
    models_available: z.array(z.string()).default(() => []),
    default_model: z.string().nullable().optional(),
    // Strukturiert seit #1458 — vorher `z.string()` mit rohem Exception-Text.
    error: StatusCheckErrorSchema.nullable().optional(),
  })
  .passthrough()

/**
 * Spiegel von `backend/app/contracts/system_status_contract.py::SystemStatusE2E`
 * (generiert nach `schemas/system-status-e2e.schema.json`).
 *
 * Meldet, ob der Backend-Prozess im E2E-Stub-Modus läuft, also ob
 * `LLMClient.chat_json` Konservenantworten statt echter Provider-Antworten
 * liefert. `stub_active` ist genau dann `true`, wenn `llm_mode === 'stub'`.
 */
export const SystemStatusE2ESchema = z
  .object({
    llm_mode: z.string().nullable().optional(),
    stub_active: z.boolean(),
  })
  .passthrough()

export const SystemStatusDiskSchema = z
  .object({
    uploads: z
      .object({
        path: z.string().optional(),
        total_bytes: z.number().nullable().optional(),
        free_bytes: z.number().nullable().optional(),
        used_pct: z.number().nullable().optional(),
        // Strukturiert seit #1458 — vorher `z.string()` mit rohem Exception-Text.
        error: StatusCheckErrorSchema.optional(),
      })
      .passthrough(),
  })
  .passthrough()

export const SystemStatusGpuSchema = z
  .object({
    nvidia_smi_available: z.boolean().optional(),
    ollama_uses_gpu: z.boolean().nullable().optional(),
    hints: z.array(z.string()).optional(),
  })
  .passthrough()

export const SystemStatusResponseSchema = z
  .object({
    backend: SystemStatusBackendSchema,
    neo4j: SystemStatusNeo4jSchema,
    ollama: SystemStatusOllamaSchema,
    // Optional, damit ein aelteres Backend ohne den Teilbaum nicht am
    // Zod-Parse scheitert. Der E2E-Helper assertiert die Anwesenheit selbst.
    e2e: SystemStatusE2ESchema.optional(),
    disk: SystemStatusDiskSchema,
    gpu: SystemStatusGpuSchema.optional(),
    timestamp: z.string(),
  })
  .passthrough()

export type SystemStatusResponse = z.infer<typeof SystemStatusResponseSchema>

File: backend/app/contracts/system_status_contract.py
"""Contracts für ``/api/status`` — SSoT für den Frontend-Zod-Spiegel.

Bislang war die Antwort von ``/api/status`` handgeschrieben und der
Zod-Spiegel in ``frontend/src/contracts/systemStatusContract.ts`` hatte keine
Gegenseite, gegen die der Schema-Drift-Gate hätte prüfen können. Dieses Modul
schließt diese Lücke zunächst für den Ollama-Teilbaum, dessen Shape sich mit
dem Provider-Gating geändert hat (``reachable`` ist jetzt dreiwertig).

Die übrigen Teilbäume (``backend``, ``neo4j``, ``disk``, ``gpu``) sind
weiterhin ungedeckt — ihre Migration gehört in einen eigenen Slice und würde
den Scope dieses Fixes sprengen. Issue #1458 strukturiert dort ausschließlich
den Fehler-Shape (``StatusCheckError``), nicht den jeweiligen Teilbaum selbst
— ``_get_neo4j_status``/``_get_disk_status`` in ``app/api/status.py`` geben
weiterhin rohe Dicts zurück, betten darin aber jetzt ``StatusCheckError``
statt eines rohen Exception-Strings ein.
"""
from __future__ import annotations

from enum import Enum

from pydantic import BaseModel, ConfigDict, Field

# Muss mit ``HttpDetectedProvider`` aus app.llm.providers.registry
# konsistent bleiben. Bewusst ``str`` statt Literal: der Wert wandert in einen
# i18n-Lookup mit Fallback, ein neuer Provider darf den Status nicht
# invalidieren.
SkippedProviderKind = str


class StatusErrorCode(str, Enum):
    """Geschlossene Fehlerklassifikation für Probe-Fehler in ``/api/status``.

    Der rohe Exception-Text (Dateipfade, Hostnamen, Treiberdetails) ist ein
    Informationsleck und für ein Frontend nicht darstellbar (Python-
    Traceback-Prosa) — er bleibt im strukturierten Log. Hier wandert nur ein
    Code, den das Frontend in einen lesbaren Satz übersetzt.

    Bewusst nicht ein Code je Exception-Klasse, sondern eine Klassifikation
    danach, wie ein Nutzer reagieren würde:

    * ``unreachable`` — Ziel (Host, Pfad) ist nicht erreichbar/vorhanden.
    * ``timeout`` — Probe hat das Zeitbudget überschritten.
    * ``auth`` — Authentifizierung/Berechtigung fehlgeschlagen.
    * ``unexpected`` — alles andere; Catch-all, damit ein neuer,
      unklassifizierter Fehler nicht crasht.
    """

    UNREACHABLE = "unreachable"
    TIMEOUT = "timeout"
    AUTH = "auth"
    UNEXPECTED = "unexpected"


class StatusCheckError(BaseModel):
    """Strukturierter Fehler-Shape für Probe-Fehler in ``/api/status``.

    Ersetzt den rohen ``str(exc)``, der bislang an drei Stellen (neo4j,
    ollama, disk) direkt in die HTTP-Antwort floss.
    """

    model_config = ConfigDict(extra="forbid")

    code: StatusErrorCode


class SystemStatusOllama(BaseModel):
    """Ollama-Teilbaum von ``/api/status``.

    ``reachable`` ist dreiwertig:

    * ``True``  — ``/api/tags`` beantwortet
    * ``False`` — Probe lief, Server nicht erreichbar (``error`` gesetzt)
    * ``None``  — Probe bewusst übersprungen, weil der aktive Provider keine
      ``/api/tags``-Route bedient (``skipped=True``, ``skipped_provider``
      gesetzt). Das ist ausdrücklich KEIN Fehlerzustand; Consumer dürfen
      ``None`` nicht als "offline" rendern.
    """

    model_config = ConfigDict(extra="forbid")

    reachable: bool | None = None
    skipped: bool = False
    # Maschinenlesbar für den i18n-Lookup im Frontend. ``reason`` bleibt
    # daneben als menschenlesbares Debug-Feld erhalten, wird aber nicht in der
    # UI gerendert — hartkodierte Backend-Prosa in der Oberfläche verstößt
    # gegen die vue-i18n-Pflicht aus AGENTS.md.
    skipped_provider: SkippedProviderKind | None = None
    reason: str | None = None
    base_url: str | None = None
    models_available: list[str] = Field(default_factory=list)
    default_model: str | None = None
    # Strukturiert seit #1458 — vorher roher ``str(exc)`` (Informationsleck,
    # fürs Frontend unrenderbarer Traceback-Text). Der Rohtext geht ins Log.
    error: StatusCheckError | None = None


class SystemStatusNeo4j(BaseModel):
    """Neo4j-Teilbaum von ``/api/status`` (Issue #1466).

    ``_get_neo4j_status`` hat drei Zweige mit unterschiedlicher Feldmenge:

    * ``storage is None`` — ``is_connected``/``last_success_ts`` werden nie
      gesetzt (kein Verbindungsversuch je unternommen).
    * Erfolgreiche Probe — alle Felder gesetzt, ``error`` ist ``None``.
    * Fehlgeschlagene Probe — alle Felder gesetzt, ``error`` strukturiert
      (``StatusCheckError``, seit #1458).

    ``_get_neo4j_status`` serialisiert deshalb mit ``exclude_unset=True``
    statt ``exclude_none=True``: nur explizit gesetzte Felder wandern in die
    Antwort, damit der erste Zweig ``is_connected``/``last_success_ts``
    weiterhin ganz auslässt statt sie als ``null`` zu senden — exakt das
    bisherige, handgeschriebene Wire-Format.
    """

    model_config = ConfigDict(extra="forbid")

    reachable: bool
    error: StatusCheckError | None = None
    uri: str | None = None
    is_connected: bool | None = None
    last_success_ts: str | None = None


class SystemStatusDiskUploads(BaseModel):
    """``uploads``-Teilbaum von ``SystemStatusDisk`` (Issue #1466).

    Im Erfolgsfall fehlt ``error`` ganz (kein ``null``-Feld); im Fehlerfall
    sind ``total_bytes``/``free_bytes``/``used_pct`` explizit ``null`` und
    ``error`` strukturiert gesetzt (seit #1458). ``_get_disk_status``
    serialisiert mit ``exclude_unset=True``, damit beide Zweige ihr
    bisheriges Wire-Format behalten.
    """

    model_config = ConfigDict(extra="forbid")

    path: str
    total_bytes: int | None = None
    free_bytes: int | None = None
    used_pct: float | None = None
    error: StatusCheckError | None = None


class SystemStatusDisk(BaseModel):
    """Disk-Teilbaum von ``/api/status`` (Issue #1466)."""

    model_config = ConfigDict(extra="forbid")

    uploads: SystemStatusDiskUploads


class SystemStatusE2E(BaseModel):
    """E2E-Harness-Teilbaum von ``/api/status``.

    Meldet, ob der **Backend-Prozess** im E2E-Stub-Modus läuft, also ob
    ``LLMClient.chat_json`` statt eines echten Providers eine Konservenantwort
    liefert (``AGORA_E2E_LLM_MODE=stub``). Der Wert stammt aus der Umgebung des
    Backend-Prozesses und wird bei jedem Request frisch gelesen.

    Warum das an der API hängt und nicht nur im Container-Log steht: die
    E2E-Suite lief bisher gegen ``process.env.AGORA_E2E_LLM_MODE`` des
    *Playwright*-Prozesses — ein Wert, der über den Zustand des Backends
    nichts aussagt. Damit konnte eine Suite grün durchlaufen, die in
    Wirklichkeit gegen einen echten Provider (oder gegen gar keinen) lief.

    Kein Geheimnis: der Teilbaum sagt nur, welchen LLM-Pfad diese Instanz
    fährt. Für Betreiber ist das eher ein Sicherheitsgewinn — eine
    Produktivinstanz, die versehentlich mit Stub-LLM läuft, produziert sonst
    still erfundene Berichte, ohne dass es irgendwo sichtbar wäre.
    """

    model_config = ConfigDict(extra="forbid")

    # Rohwert von ``AGORA_E2E_LLM_MODE``; ``None``, wenn die Variable im
    # Backend-Prozess nicht gesetzt ist (Normalfall in Produktion).
    llm_mode: str | None = None
    # Abgeleitet: genau dann ``True``, wenn ``llm_mode == "stub"``. Der
    # E2E-Helper assertiert hierauf, damit er sich nicht auf String-Vergleiche
    # im Testcode verlassen muss.
    stub_active: bool = False
